### PR TITLE
Adds multi contributor functions to the Investor Staking Contracts

### PR DESCRIPTION
**Pull request by darcys22**

### CODE REVIEW

File: contracts/utils/TokenVestingStaking.sol [hunk 1]
```diff
@@ -146,6 +168,51 @@ contract TokenVestingStaking is ITokenVestingStaking, Shared {
...
+    /**
+     * @notice Withdraws contribution from a ServiceNodeContribution contract
+     * @param contributionContractAddress The address of the ServiceNodeContribution contract
+     */
+    function withdrawContribution(address contributionContractAddress) external onlyBeneficiary notRevoked afterStart {
+        require(contributionFactory.isContractDeployed(contributionContractAddress), "Invalid contribution contract");
+        IServiceNodeContribution contributionContract = IServiceNodeContribution(contributionContractAddress);
+        bool foundNode = false;
+        for (uint256 i = 1; i < investorServiceNodes.length + 1; i++) {
+            if (investorServiceNodes[i - 1].contributionContract == contributionContractAddress) {
```
Comment (Doy-lee) [inline contracts/utils/TokenVestingStaking.sol:205]: Similar stale node problem here if the contract is reset/cancelled or reused. The entry will persist and there won't be a way to get rid of it. I think we can just flush them out on claimRewards, or all these functions that interact with the contract have a modifier to groom the list.

**Review by Doy-lee — CHANGES_REQUESTED** [review]

File: contracts/utils/TokenVestingStaking.sol [hunk 2]
```diff
@@ -119,11 +127,25 @@ contract TokenVestingStaking is ITokenVestingStaking, Shared {
...
         for (uint256 i = 1; i < length + 1; i++) {
-            IServiceNodeRewards.ServiceNode memory sn = stakingRewardsContract.serviceNodes(
-                investorServiceNodes[i - 1].serviceNodeID
-            );
+            ServiceNode memory investorNode = investorServiceNodes[i - 1];
+            uint64 serviceNodeID = investorNode.serviceNodeID;
+            IServiceNodeContribution contributionContract = IServiceNodeContribution(investorNode.contributionContract);
+            if (address(contributionContract) != address(0)) {
+                uint256 contributedAmount = contributionContract.contributions(address(this));
+                if (!contributionContract.finalized() && contributedAmount > 0)
```
[the remark this answers is not in the document]
Comment (darcys22) [reply contracts/utils/TokenVestingStaking.sol:135]: So contributedAmount here is ment to be checking if the contributions contract has been reset, so the statement here is saying 
```
if ( not finalised and not reset)
    continue // no more action needed to update our internal state
```
If it does find it finalised and reset then it goes and finds the bls key to check the staking rewards contract.

I think you are correct here with the `contribute -> reset -> contribute -> finalise` causing issues. The `contributeFunds()` function should have some more logic to handle that in the second contribute. I'm thinking all contributeFunds calls should synchronize the invesor contract with the contribution contract.

File: contracts/utils/TokenVestingStaking.sol [hunk 3]
```diff
@@ -119,11 +127,25 @@ contract TokenVestingStaking is ITokenVestingStaking, Shared {
         uint256 unstaked = 0;
         uint256 length = investorServiceNodes.length;
         for (uint256 i = 1; i < length + 1; i++) {
-            IServiceNodeRewards.ServiceNode memory sn = stakingRewardsContract.serviceNodes(
-                investorServiceNodes[i - 1].serviceNodeID
-            );
+            ServiceNode memory investorNode = investorServiceNodes[i - 1];
+            uint64 serviceNodeID = investorNode.serviceNodeID;
```
[the remark this answers is not in the document]
Comment (darcys22) [reply contracts/utils/TokenVestingStaking.sol:131]: I think there might be a way to solve this by checking and syncing the investor contract whenever the investor contributes.

**Comment by Doy-lee** [comment]
Issues have been addressed and merged in via https://github.com/oxen-io/eth-sn-contracts/pull/87

**Pull request closed without merging**